typed_syscall_data/README.md

```diff
@@ -0,0 +1,481 @@
+
+An ARM, the system call being invoked is passed via the `svc` instruction; on
+RISC-V it is passed via `x10` (this is why `x10` is a return register but not
+an argument register).
+
+Convention: In this document I will often list a range of registers, such as
+r1-r15. Whenever such a range is specified, it means the list of registers that
+are in that range and exist on the platform a particular Tock system is running
+on. For example, r10-r13 refers to (x6, x7, x28, x29) on RISC-V but only (v7,
+v8) on ARM because r12 and r13 do not exist on ARM.
+
+## Proposed Syscall ABI
+
+This section proposes a system call ABI for Tock. This system call ABI is based
+on [TRD
+104](https://github.com/tock/tock/blob/master/doc/reference/trd104-syscalls.md)
+and omits many details that are either the same as TRD 104 or which are
+irrelevant to exploring the concept of typed system calls.
+
+### Return Values
+
+All system calls return an `ArbitraryData(N)`, where `N` is the maximum of:
+
+1. The number of registers required to return the Success variant.
+1. The number of registers required to return the Failure variant.
+1. The number of registers required to return a `(ErrorCode)`.
+
+Note: All system calls must specify their success and failure variant, so that
+userspace libraries can determine how many registers may be clobbered by the
+syscall.
+
+If userspace tries to invoke a system call that the kernel does not recognize or
+does not support, or a system call on a system call driver that does not exist,
+the kernel will return type `(ErrorCode)`.
+
+### Upcall arguments
+
+Semantically, upcalls have two arguments: the userdata pointer (which was passed
+to Subscribe), and an `ArbitraryData(4)` of data from the syscall driver. This
+allows syscall drivers to pass up to three values to each invoked upcall. Upcall
+functions must have an ABI compatible with the following signature (specified in
+both Rust and C):
+
+```C
+struct UpcallArbitraryData {
+    Register registers[4];
+}
+
+typedef void(void*, UpcallArbitraryData) UpcallFn;
+```
+
+```Rust
+#[repr(C)]
+struct UpcallArbitraryData {
+    registers: [Register; 4],
+}
+
+type UpcallFn = unsafe extern "C" fn(*mut (), UpcallArbitraryData);
+```
+
+### Yield
+
+Arguments:
+
+| Yield type | a1  | a2            | a3               |
+| ---------- | --- | ------------- | ---------------- |
+| no-wait    | `0` | Unused        | Unused           |
+| wait       | `1` | Unused        | Unused           |
+| wait-for   | `2` | Driver number | Subscribe number |
+
+If there is no pending upcall, yield-no-wait will return `()`.
+
+If there is an upcall to invoke, yield-wait and yield-no-wait return `(upcall fn
+pointer, userdata: *mut (), Register, Register, Register, Register)`. The
+userspace Yield function should then pack the register values into an
+`UpcallArbitraryData` and invoke the function pointer with the userdata pointer
+and `UpcallArbitraryData` as parameters.
```

Comment:
This is new, right? I thought Tock actually put a call onto the user's stack, but maybe it doesn't and I just haven't looked in a long time.

Reply:
Yes, the issue is that ARM's C ABI has only four argument registers but the upcall callback needs five registers' worth of arguments. I added the following paragraph to explain the change:

> Note: Unlike TRD 104, in this syscall ABI Yield returns the upcall's arguments to the userspace library and expects the userspace library to invoke the callback. This allows for more than four registers' worth of arguments to be passed to the upcall without requiring the kernel to push arguments onto the userspace stack.